### PR TITLE
[test-runner] Try to dump threads on timeout with lldb and gdb + [mini] Factor duplicates mono_gdb_render_native_backtraces

### PR DESCRIPTION
**Pull request by ludovic-henry**

### CODE REVIEW

**Comment by ludovic-henry** [comment]
cc @vargaz @akoeplinger

File: mono/tests/test-runner.cs [hunk 1]
```diff
@@ -35,6 +35,7 @@ public class TestRunner
 	class ProcessData {
 		public string test;
 		public StringBuilder stdout, stderr;
+		public object stdout_lock = new object (), stderr_lock = new object ();
```
Comment (akoeplinger) [inline mono/tests/test-runner.cs]: nit: stdoutLock would be the proper C# naming

File: mono/tests/test-runner.cs [hunk 2]
```diff
@@ -611,4 +608,119 @@ class TestInfo {
...
+		{
+			sw.WriteLine ("attach " + pid);
+			sw.WriteLine ("info threads");
+			sw.WriteLine ("thread apply all p mono_print_thread_dump(0)");
+			sw.WriteLine ("thread apply all backtrace");
+			sw.Flush ();
+
+			ProcessStartInfo psi = new ProcessStartInfo {
+				FileName = "gdb",
+				Arguments = "-batch -x " + filename + " -nx",
```
Comment (akoeplinger) [inline mono/tests/test-runner.cs]: Same here.

File: mono/tests/test-runner.cs [hunk 3]
```diff
@@ -611,4 +608,119 @@ class TestInfo {
...
+			sw.WriteLine ("process attach --pid " + pid);
+			sw.WriteLine ("thread list");
+			sw.WriteLine ("thread backtrace all");
+			sw.WriteLine ("detach");
+			sw.WriteLine ("quit");
+			sw.Flush ();
+
+			ProcessStartInfo psi = new ProcessStartInfo {
+				FileName = "lldb",
+				Arguments = "--batch --source " + filename + " --no-lldbinit",
```
Comment (akoeplinger) [inline mono/tests/test-runner.cs]: I assume filename could contain spaces e.g. on Windows so maybe better to enclose in quotes.

File: mono/tests/test-runner.cs [hunk 4]
```diff
@@ -611,4 +608,119 @@ class TestInfo {
 		// Char ::= #x9 | #xA | #xD | [#x20-#xD7FF] | [#xE000-#xFFFD] | [#x10000-#x10FFFF]	/* any Unicode character, excluding the surrogate blocks, FFFE, and FFFF. */
 		return Regex.Replace (text, @"[^\x09\x0A\x0D\x20-\uD7FF\uE000-\uFFFD\u10000-\u10FFFF]", "");
 	}
+
+	static void TryThreadDump (int pid, ProcessData data)
+	{
+		try {
+			TryGDB (pid, data);
+			return;
+		} catch (Exception) {
```
Comment (akoeplinger) [inline mono/tests/test-runner.cs]: just `catch { }` since we don't do anything with Exception anyway.

File: mono/tests/test-runner.cs (hunk 1 above)
Comment (vargaz) [inline mono/tests/test-runner.cs]: We use the glib/mono naming convention in c# too, so either is ok.

**Pull request merged**